libcloud/compute/drivers/cloudstack.py

```diff
@@ -654,6 +654,12 @@ def list_nodes(self, project=None):
             args['projectid'] = project.id
         vms = self._sync_request('listVirtualMachines', params=args)
         addrs = self._sync_request('listPublicIpAddresses', params=args)
+        port_forwarding_rules = self._sync_request('listPortForwardingRules')
```

Comment:
You could dispense with a lot of the `.get(key, [])` complexity if you'd use a defaultdict here.

```python
port_forwarding_rules = defaultdict({}, **self._sync_request('listPortForwardingRules)
```

While a little more complex to set up here, this will make the rest of the code much easier to read.